### ISSUE TITLE
[Experiment] Incr. Comp.: Replace query result hashing with early-exit deserialization.

### ISSUE THREAD

**Issue by michaelwoerister** [rust-lang/rust#58155]
The red-green algorithm of incremental compilation needs to compare query results in order to determine if a result has actually changed after it is has been recomputed. At the moment this is facilitated by creating a `Fingerprint` of every result and storing that `Fingerprint` together with the `DepGraph`. This fingerprinting takes up a considerable amount of time during the initial compilation session (and any other session that re-computes many results). 

There is a different way of implementing this check that might be more efficient: Instead of computing fingerprints, we "just" store the result in the on-disk cache. When we encounter the need to compare a cached result to the newly computed one, we try to load it from the cache. However, we provide an early exit path to this deserialization: As soon as the deserializer encounters something that is not yet in an interner, it fails gracefully. Since any part of the result already has to be present in the interners if it were actually equal to the re-computed one, we know that it must be different. This early exit also has the side effect that we'll never load any unneeded data into interners. 

The advantages of this approach are:
- fingerprinting is eager: We always have to do it although we might never need the fingerprint. Interner-based comparison postpones most of the work to the point when it's actually needed.
- fingerprinting is eager in a second way: It always has to process the whole value.
- fingerprinting has to duplicate work unless there is an explicit sharing mechanism (e.g. caching the hash of sub-components and then hashing those hashes) - interner-based comparison has sharing built right into it. 
- fingerprinting has to use a good quality hash function (SipHash at the moment). Interners use FxHash which is a lot faster.
- because most of the cached results contain interned values, much of them has to be hashed during deserialization anyway. 

The downside of this approach is: 
- every query result has to be stored in the on-disk cache, so it's available for comparison. 

The steps to implement this are:
1. Make sure that all query results are stored to the on-disk cache.
2. Allow for deserialization to fail if something is not in the interner.
3. Make red-green marking use the new approach.
4. Turn off fingerprinting for query results.

(TODO: links to code)

cc @Xanewok @rust-lang/wg-compiler-performance

**Comment by michaelwoerister** [comment]
This is where query results are stored to the on-disk cache:
https://github.com/rust-lang/rust/blob/2596bc1368d1e3d34c9a7841ad87a3100f01cbad/src/librustc/ty/query/on_disk_cache.rs#L204-L248

Since almost all query results have to be serialized now, it would probably make sense to implement this in the query definition macro in `plumbing.rs` (or something that walks all query tables).

**Comment by michaelwoerister** [comment]
This is where the red-green system does the check whether something is red or green:
https://github.com/rust-lang/rust/blob/2596bc1368d1e3d34c9a7841ad87a3100f01cbad/src/librustc/dep_graph/graph.rs#L284-L299

**Comment by michaelwoerister** [comment]
This is where types are created:
https://github.com/rust-lang/rust/blob/2efa31b2d9bf171fecd294b8e0126d8ffdb453e3/src/librustc/ty/context.rs#L2454-L2457

There would have to be an alternative method that fails if the thing is not in the interner yet. Ideally `intern_ty` is updated to support both options:
https://github.com/rust-lang/rust/blob/2efa31b2d9bf171fecd294b8e0126d8ffdb453e3/src/librustc/ty/context.rs#L152-L198
That would have to be benchmarked though. We don't want to regress performance in the common case.

**Comment by michaelwoerister** [comment]
I did a quick (and messy) proof-of-concept implementation of this here: 
https://github.com/michaelwoerister/rust/tree/no-hash-experiment

My findings are:
- It would be lots and lots of work to entirely get rid of stable hashing in the compiler. The POC has to fall back to fingerprint comparison for anything that is not in the cache and it is unlikely that we would ever want to store *everything* in the cache.
- A special case of the above are things from upstream crate metadata. There is no reason to ever store these in the cache. However, supporting the new comparison strategy for upstream crate metadata comes with a whole new set of implementation headaches (at least in the current version of the query framework).
- With the POC I ran into trouble where `Eq` would treat the new and the equal deserialized value as unequal. Unless there is a bug somewhere else, I suspect that values are not always entirely equal according to `Eq` -- we'd probably need a `StableEq` that like `HashStable` that ignores unimportant details. But adding such a trait would not be a win over just sticking to `HashStable`. 
- Using fingerprints is actually very nice in some respects. They are all the same size on disk, it is easy to deserialize them in different contexts, etc.

For these reasons I think it does not make sense to pursue this further at the moment.
